src/main/scala/firrtl/passes/wiring/Wiring.scala

```diff
@@ -10,157 +10,160 @@ import firrtl.Mappers._
 import scala.collection.mutable
 import firrtl.annotations._
 import firrtl.annotations.AnnotationUtils._
+import firrtl.analyses.InstanceGraph
 import WiringUtils._
 
 case class WiringException(msg: String) extends PassException(msg)
 
-case class WiringInfo(source: String, comp: String, sinks: Set[String], pin: String, top: String)
+case class WiringInfo(source: Named, sinks: Seq[Named], pin: String)
 
 class Wiring(wiSeq: Seq[WiringInfo]) extends Pass {
   def run(c: Circuit): Circuit = {
     wiSeq.foldLeft(c) { (circuit, wi) => wire(circuit, wi) }
   }
 
-  /** Add pins to modules and wires a signal to them, under the scope of a specified top module
-    * Description:
-    *   Adds a pin to each sink module
-    *   Punches ports up from the source signal to the specified top module
-    *   Punches ports down to each sink module
-    *   Wires the source up and down, connecting to all sink modules
-    * Restrictions:
-    *   - Can only have one source module instance under scope of the specified top
-    *   - All instances of each sink module must be under the scope of the specified top
-    * Notes:
-    *   - No module uniquification occurs (due to imposed restrictions)
+  /** Wires Sources to Sinks
+    *
+    * Sinks are wired to their closest source through their lowest
+    * common ancestor (LCA). Verbosely, this modifies the circuit in
+    * the following ways:
+    *   - Adds a pin to each sink module
+    *   - Punches ports up from source signals to the LCA of each sink
+    *   - Punches ports down from LCAs to each sink module
+    *   - Wires sources up to LCA, sinks down from LCA, and across each LCA
+    *
+    * @param c The circuit to modify
+    * @param wi A WiringInfo object containing source, sink, component, and pin information
+    * @return The modified circuit
+    *
+    * @throws WiringException if a sink is equidistant to two sources
```

Comment:
Please correct me if I am wrong but it feels like this refactor is intending to do all wirings in one fell swoop, but it doesn't look like that's happening. This exception for example mentions a sink being equidistant to two sources, but given that we call `wire` for each `WiringInfo`, how can there be two sources for a single sink? Don't we recreate all of these data structures for each wiring?

Reply:
This is currently doing all wirings for a given `Source Component -> Set(Sink Component)` mapping in one fell swoop. I would much prefer if this was doing _all_ wirings for all wiring annotations at once (I intended that specifically to be a future addition). That's a little more complicated, but not overly so. That will improve the performance, and justify the whole Euler Tour nonsense better, as you only pay that construction cost once for all wiring annotations (unless you have passes that generate new wiring annotations and call the wiring transform...).

This problem of a "sink being equidistant to two sources" comes up because components are tied to a module and not an instance. Hence, some weird things can happen once you start instantiating multiple copies of the same source module. 

Specifically, you can create a circuit like the following test where you have two sources at an equivalent level in the hierarchy that are supposed to wire to a single sink that is at a higher level in the hierarchy. 
  - Equidistant problem:  https://github.com/seldridge/firrtl/blob/2d7f86f14f7270a43cdeed0dd0c33e1a0d00fc4a/src/test/scala/firrtlTests/WiringTests.scala#L480
Here, there's no clear way to determine what the user intended ("Wire these modules", but the user has specified a hierarchy that doesn't make sense). I do not expect most people to hit this, but it can come up if you were trying to, e.g., wire a rocket core to an L2. This would work fine for a single core, but then you'd run into a problem if you instantiated two cores. I'd say that if you really want this type of behavior, you need to be using diplomacy or some library that can figure out what you instantiated and then generate the logic you need.

Alternatively, this could be represented as "Unable to determine unique mapping of source to sink".

Seem reasonable?

Alternatively, the much more frequent case is the following where the sink is encased neatly under the source (though this is really just a special case of it not being equidistant). Here, it's unambiguous what the user meant
  - No Equidistant Problem: https://github.com/seldridge/firrtl/blob/2d7f86f14f7270a43cdeed0dd0c33e1a0d00fc4a/src/test/scala/firrtlTests/WiringTests.scala#L435

Reply:
I think it is totally reasonable to do all wirings in one fell swoop in a future PR. Thank you for the clarifications! A++ for having tests to illustrate!